src/power_grid_model/data_types.py

```diff
@@ -13,60 +13,84 @@
 
 # When we're dropping python 3.8, we should introduce proper NumPy type hinting
 
-SparseBatchArray = Dict[str, np.ndarray]
+SingleArray = Union[np.ndarray]
 """
-A sparse batch array is a dictionary containing the keys `indptr` and `data`.
+A single array is a one-dimensional structured containing a list of components of the same type.
+
+- Examples:
 
-- indptr: a one-dimensional numpy int64 array
-- data: a one-dimensional structured numpy array. The exact dtype depends on the type of component.
-- Example: {"indptr": <1d-array>, "data": <1d-array>}
+    - structure: <1d-array>
+    - concrete: array([(0, 10500.0), (0, 10500.0)], dtype=power_grid_meta_data["input"]["node"].dtype)
 """
 
-BatchArray = Union[np.ndarray, SparseBatchArray]
+DenseBatchArray = Union[np.ndarray]
+"""
+A dense batch array is a two-dimensional structured numpy array containing a list of components of 
+the same type for each scenario.
 """
-A batch is a either a dense or a sparse batch array.
+
+SparseBatchArray = Dict[str, Union[np.ndarray, SingleArray]]
+"""
+A sparse batch array is a dictionary containing the keys `indptr` and `data`.
+
+- data: a :class:`SingleArray`. The exact dtype depends on the type of component.
+- indptr: a one-dimensional numpy int64 array containing n+1 elements where n is the amount of scenarios.
+
+    - The elements are the indices in the data that point to the first element of that scenario.
+    - The last element is one after the data index of the last element of the last scenario.
+    - Usually, the last element will therefore be the size of the data.
 
 - Examples:
 
-dense: <2d-array>
+    - structure: {"indptr": <1d-array>, "data": :class:`SingleArray`}
+    - concrete example: {"indptr": [0, 2, 2, 3], "data": [(0, 1, 1), (1, 1, 1), (0, 0, 0)]}
+
+        - the scenario 0 sets the statuses of components ids 0 and 1 to 1 (and keeps defaults for other components)
+        - scenario 1 keeps the default values for all components
+        - scenario 2 sets the statuses of component with id 0 to 0 (and keeps defaults for other components)
+"""
+
+BatchArray = Union[DenseBatchArray, SparseBatchArray]
+"""
+A batch array is a either a :class:`DenseBatchArray` or a :class:`SparseBatchArray`.
+"""
 
-sparse: {"indptr": <1d-array>, "data": <1d-array>}
+DataArray = Union[SingleArray, BatchArray]
+"""
+A data array can be a :class:`SingleArray` or a :class:`BatchArray`.
 """
 
-SingleDataset = Dict[str, np.ndarray]
+SingleDataset = Union[SingleArray]
```

Comment:
Is this equivalent to Dict[str, ndarray]? This data type is being used externally too.

Reply:
whoops, was a booboo